File: src/evosim/node.ts
import Edge from "./edge";
import Physics from "./physics";
import Vector from "./vector";
import {v4 as uuid} from "uuid";

type NodeState = {
	position: Vector;
	velocity: Vector;
	grounded: boolean;
}

export default class Node {
	public id: string;
	public radius: number;
	public mass: number;
	public friction: number;
	public position: Vector;
	public edges: Set<Edge>;

	public state?: NodeState;

	constructor(radius: number, friction: number, position: Vector) {
		this.id = uuid();
		this.radius = radius;
		this.mass = Math.PI * Math.pow(this.radius, 2);
		this.friction = friction;
		this.position = position;
		this.edges = new Set();
	}

	get acceleration(): Vector {
		const state = this.state;
		if (!state) throw Error();

		let force = Physics.gravity;
		this.edges.forEach(edge => force.add(edge.getForce(this.id)));

		if (state.grounded) {
			const friction = this.friction * Math.abs(force.y);
			force.x -= friction * Math.sign(state.velocity.x);
		}

		return force.div(this.mass);
	}

	start() {
		this.state = {
			position: this.position.clone(),
			velocity: new Vector(),
			grounded: false
		};
	}

	update(dt: number): void {
		let state = this.state;
		if (!state) throw Error();

		let acceleration = this.acceleration;
		state.velocity.add(acceleration.over(dt));
		state.position.add(state.velocity.over(dt));

		state.grounded = state.position.y < this.radius + 0.02;
		if (state.position.y < this.radius) {
			state.position.y = this.radius;
			state.velocity.y = Math.max(0, state.velocity.y);
		}
		state.velocity.mult(0.97);
	}
}

File: src-tauri/src/evosim/node.rs
use std::sync::{Arc, RwLock};

use super::{edge::EdgeState, physics, vector::Vector};

#[derive(Clone, Copy)]
pub struct Node {
	pub id: usize,
	pub radius: f64,
	pub mass: f64,
	pub friction: f64,
	pub position: Vector,
}

pub struct NodeState {
	pub node: Node,
	pub position: Vector,
	pub velocity: Vector,
	pub edges: Vec<Arc<RwLock<EdgeState>>>,
	pub grounded: bool
}

impl Node {
	pub fn new(
		id: usize,
		radius: f64,
		friction: f64,
		position: Vector
	) -> Self {
		Node {
			id,
			radius,
			mass: 3.14159 * radius.powi(2),
			friction,
			position
		}
	}
}

impl NodeState {
	pub fn new(node: Node) -> Self {
		NodeState {
			node,
			position: node.position.clone(),
			velocity: Vector::blank(),
			edges: Vec::new(),
			grounded: false
		}
	}

	pub fn acceleration(&mut self) -> Vector {
		let mut force = physics::GRAVITY.clone();
		for edge in &self.edges {
			force += edge.read().unwrap().get_force(self.node.id);
		}

		if self.grounded {
			let friction = self.node.friction * force.y.abs();
			force.x -= friction * self.velocity.x.signum();
		}

		force / self.node.mass
	}

	pub fn tick(&mut self, dt: f64) {
		let acceleration = self.acceleration();
		self.velocity += acceleration * dt;
		self.position += self.velocity * dt;

		self.grounded = self.position.y < self.node.radius + 0.02;
		if self.position.y < self.node.radius {
			self.position.y = self.node.radius;
			self.velocity.y = self.velocity.y.max(0.0);
		}
		self.velocity *= 0.97;
	}
}
